File: TDD/TDD Coding Practice String Transformer/Code/Java/src/fayelab/tdd/stringtransformer/addall/Presenter.java
package fayelab.tdd.stringtransformer.addall;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class Presenter
{
    private View view;
    private BusinessLogic businessLogic;
    
    private List<String> chainTransIds = new ArrayList<>();
    private List<String> availableTransIds;

    public Presenter(View view, BusinessLogic businessLogic)
    {
        this.view = view;
        this.businessLogic = businessLogic;
        
        this.view.setPresenter(this);
    }

    public void init()
    {
        availableTransIds = businessLogic.getAllTransIds();
        view.presentAvailableTransIds(availableTransIds);
    }

    public void addTransformer()
    {
        String selectedAvailableTransId = view.getSelectedAvailableTransId();
        if(notExistedInChain(selectedAvailableTransId))
        {
            chainTransIds.add(selectedAvailableTransId);
        }
        
        view.presentChainTransIds(chainTransIds);
    }

    private boolean notExistedInChain(String transId)
    {
        return !chainTransIds.contains(transId);
    }

    public void removeTransformer()
    {
        chainTransIds.remove(view.getSelectedChainTransId());
        view.presentChainTransIds(chainTransIds);
    }

    public void applyTransformerChain()
    {
        String sourceStr = view.getSourceStr();
        if(!checkInput(sourceStr, chainTransIds))
        {
            return;
        }
        
        view.presentResultStr(businessLogic.transform(sourceStr, chainTransIds));
    }

    public void removeAllTransformers()
    {
        chainTransIds.clear();
        view.presentChainTransIds(chainTransIds);
    }

    public void addAllTransformers()
    {
        chainTransIds.clear();
        chainTransIds.addAll(availableTransIds);
        view.presentChainTransIds(chainTransIds);
    }

    private boolean checkInput(String sourceStr, List<String> chain)
    {
        return checkSourceStr(sourceStr) && checkChain(chain);
    }

    private boolean checkSourceStr(String str)
    {
        return checkParam(str, s -> s.isEmpty(), () -> view.onEmptySourceStrInput());
    }

    private boolean checkChain(List<String> chain)
    {
        return checkParam(chain, c -> c.isEmpty(), () -> view.onEmptyChainInput());
    }
    
    private <T> boolean checkParam(T param, Predicate<T> pred, Runnable notify)
    {
        if(pred.test(param))
        {
            notify.run();
            return false;
        }
        
        return true;
    }
}

File: TDD/TDD Coding Practice String Transformer/Code/Python/addall/presenter.py
class Presenter(object):
    def __init__(self, view, businesslogic):
        self.view = view
        self.businesslogic = businesslogic
        self.view.set_presenter(self)

        self.chain_transids = []
        self.avail_transids = None

    def init(self):
        self.avail_transids = self.businesslogic.get_all_transids()
        self.view.present_available_transids(self.avail_transids)

    def add_transformer(self):
        selected_transid = self.view.get_selected_available_transid()
        if self.not_exist_in_chain(selected_transid):
            self.chain_transids.append(selected_transid)
        self.view.present_chain_transids(self.chain_transids)

    def remove_transformer(self):
        self.chain_transids.remove(self.view.get_selected_chain_transid())
        self.view.present_chain_transids(self.chain_transids)

    def apply_transformer_chain(self):
        sourcestr = self.view.get_sourcestr()

        if not self.check_input(sourcestr, self.chain_transids):
            return

        self.view.present_resultstr(
            self.businesslogic.transform(sourcestr, self.chain_transids))

    def remove_all_transformers(self):
        del self.chain_transids[:]
        self.view.present_chain_transids(self.chain_transids)

    def add_all_transformers(self):
        del self.chain_transids[:]
        self.chain_transids.extend(self.avail_transids)
        self.view.present_chain_transids(self.chain_transids)

    def not_exist_in_chain(self, transid):
        return transid not in self.chain_transids

    def check_input(self, sourcestr, _chain_transids):
        paramrules = [
            (sourcestr, lambda s: s == '', self.view.on_empty_sourcestr_input),
            (_chain_transids, lambda lst: lst == [], self.view.on_empty_chain_input)
        ]

        def check_param(param, predication, action):
            if predication(param):
                action()
                return False
            return True

        for paramrule in paramrules:
            if not check_param(paramrule[0], paramrule[1], paramrule[2]):
                return False

        return True